book/src/pong-tutorial/pong-tutorial-04.md

```diff
@@ -368,7 +368,123 @@ let game_data = GameDataBuilder::default()
+# use amethyst::prelude::*;
+# struct MyState;
+# impl SimpleState for MyState {
+fn update(&mut self, data: &mut StateData<'_, GameData<'_, '_>>) -> SimpleTrans {
+    Trans::None
+}
+# }
+```
+
+That method allows you to transition out of state,
+and we don't want to do that. For that reason we will always return `Trans::None`,
+which tells it to not transition.
+
+Now we have to move paddle creation to that method, and add some delay to it. Our `update` runs every frame,
+so in order to do something only once after given time, we have to use our local state.
+Additionaly, notice that `initialise_paddles` requires us to provide the `sprite_sheet_handle`, but it was created
+as a local variable inside `on_start`. For that reason we have to make it a part of the state too.
+
+Let's add some fields to our `Pong` struct:
+
+```rust,edition2018,no_run,noplaypen
+#[derive(Default)]
+pub struct Pong {
+    ball_spawn_timer: Option<f32>,
+    sprite_sheet_handle: Option<Handle<SpriteSheet>>,
+}
+```
+
+Our timer is represented by `Option<f32>`, which will count down to zero when available, and be replaced with `None` after
+the time has passed. Our sprite sheet handle is also inside `Option`, because we can't create it inside `Pong` constructor.
+It will be created inside `on_start` method instead.
+
+We've also added `#[derive(Default)]`, which will automatically implement `Default` trait for us, which allows to create
+default empty state. Now let's use that inside our `Application` creation code in `main.rs`:
+
+```rust,edition2018,no_run,noplaypen
+# extern crate amethyst;
+# use amethyst::prelude::*;
+#
+# struct Pong;
+# impl SimpleState for Pong { }
+# fn main() -> amethyst::Result<()> {
+#   let game_data = GameDataBuilder::default();
+#   let assets_dir = "/";
+let mut game = Application::new(assets_dir, Pong::default(), game_data)?;
+#   Ok(())
+# }
+```
+
+Now let's finish our timer and ball spawning code. We have to do two things:
+- First, we have to initialize our state and remove `initialise_ball` from `on_start`,
+- then we have to `initialise_ball` once after the time has passed inside `update`:
+
+```rust,edition2018,no_run,noplaypen
+
+# pub struct Pong {
+#     ball_spawn_timer: Option<f32>,
+#     sprite_sheet_handle: Option<Handle<SpriteSheet>>,
+# }
+impl SimpleState for Pong {
+    fn on_start(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        let world = data.world;
+
+        // Wait one second before spawning the ball.
+        self.ball_spawn_timer.replace(1.0);
+
+        // Load the spritesheet necessary to render the graphics.
+        // `spritesheet` is the layout of the sprites on the image;
+        // `texture` is the pixel data.
+        self.sprite_sheet_handle.replace(load_sprite_sheet(world));
+        initialise_paddles(world, self.sprite_sheet_handle.clone().unwrap());
+        initialise_camera(world);
+    }
+
+    fn update(&mut self, data: &mut StateData<'_, GameData<'_, '_>>) -> SimpleTrans {
```

Comment:
I'm surprised that function needs to be so convoluted? Doesn't NLL take care of most of that?

Reply:
Unfortunately no. In this case I've got the error without the extra scope.